### PR TITLE
Add SetOf builtin contract

### PR DESCRIPTION
**Pull request by waterlink**
To me this contract looks generic enough to be included as builtin contract. What do you think?

### CODE REVIEW

File: lib/contracts/builtin_contracts.rb [hunk 1]
```diff
@@ -259,6 +259,29 @@ def to_s
...
 
+  # Takes a contract. The related argument must be a set.
+  # Checks the contract against every element of the set.
+  # If it passes for all elements, the contract passes.
+  # Example: <tt>SetOf[Num]</tt>
+  class SetOf < ArrayOf
+    def valid?(vals)
+      lazy_require_set
+      return false unless vals.is_a?(Set)
+      super(vals.to_a)
```
Comment (egonSchiele) [inline lib/contracts/builtin_contracts.rb]: Hidden memory allocation here with `to_a`...I would just use `vals.each`.

**Comment by egonSchiele** [comment]
Very nice idea, just don't like the hidden performance/memory penalty^

**Comment by alex-fedorov** [comment]
@egonSchiele removed memory penalty by extracting common abstraction to private `CollectionOf` contract and made `ArrayOf` and `SetOf` just specific factories of `CollectionOf` contract

**Comment by waterlink** [comment]
Rebased against green master, and this one should become green too

**Pull request merged**